File: python/src/pedro_agentware/llmcontext/strategies.py
"""Compaction strategies for context management."""

from dataclasses import dataclass
from typing import Protocol

from pedro_agentware.llm import Message
from pedro_agentware.middleware.types import MessageType


class TokenCounter(Protocol):
    """Protocol for token counting functions."""

    def __call__(self, messages: list[Message]) -> int:
        """Count tokens in messages."""
        ...


class CompactionStrategy(Protocol):
    """Protocol for compaction strategies."""

    def compact(
        self, messages: list[Message], target_tokens: int, counter: TokenCounter
    ) -> list[Message]:
        """Compact messages to fit within target token count."""
        ...

    def name(self) -> str:
        """Return the strategy name."""
        ...


def _is_nudge_type(msg_type: MessageType) -> bool:
    return (
        msg_type == MessageType.STEP_NUDGE
        or msg_type == MessageType.RETRY_NUDGE
        or msg_type == MessageType.PREREQUISITE_NUDGE
    )


@dataclass
class TieredCompact:
    """3-phase compaction strategy that prunes messages in semantic priority order.

    Compaction Priority (cut first -> preserve last):
    1. STEP_NUDGE, RETRY_NUDGE, PREREQUISITE_NUDGE
    2. TOOL_RESULT - truncate to 200 chars (Phase 1)
    3. TOOL_RESULT - drop entirely (Phase 2)
    4. REASONING, TEXT_RESPONSE - drop (Phase 3)

    messages[0] and messages[1] are always protected.
    """

    keep_recent: int = 2
    phase_thresholds: tuple[float, float, float] | None = None
    truncate_chars: int = 200

    def __post_init__(self) -> None:
        if self.phase_thresholds is None:
            self.phase_thresholds = (0.75, 0.75, 0.75)

    def name(self) -> str:
        return "TieredCompact"

    def compact(
        self, messages: list[Message], target_tokens: int, counter: TokenCounter
    ) -> list[Message]:
        """Compact messages using 3-phase tiered strategy."""
        if not messages:
            return []

        result = list(messages)

        eligible_end = _find_eligible_end(result, self.keep_recent)

        current_tokens = counter(result)

        if current_tokens <= target_tokens:
            return result

        result = self._phase1_compact(result, eligible_end, counter)
        current_tokens = counter(result)
        if current_tokens <= target_tokens:
            return result

        result = self._phase2_compact(result, eligible_end, counter)
        current_tokens = counter(result)
        if current_tokens <= target_tokens:
            return result

        result = self._phase3_compact(result, eligible_end, counter)
        return result

    def _protected_steps(self, messages: list[Message]) -> dict[int, bool]:
        """Get set of protected step indices."""
        protected: dict[int, bool] = {}

        if self.keep_recent <= 0:
            return protected

        steps: list[int] = []
        step_set: set[int] = set()
        for m in messages:
            if m.meta.step_index is not None and m.meta.step_index not in step_set:
                steps.append(m.meta.step_index)
                step_set.add(m.meta.step_index)

        if not steps:
            return protected

        if self.keep_recent >= len(steps):
            for s in steps:
                protected[s] = True
            return protected

        start_idx = len(steps) - self.keep_recent
        for i in range(start_idx, len(steps)):
            protected[steps[i]] = True

        return protected

    def _is_protected(self, msg: Message, protected_steps: dict[int, bool]) -> bool:
        """Check if message is in a protected step."""
        if msg.meta.step_index is None:
            return False
        return protected_steps.get(msg.meta.step_index, False)

    def _phase1_compact(
        self, messages: list[Message], eligible_end: int, counter: TokenCounter
    ) -> list[Message]:
        """Phase 1: drop nudges, truncate TOOL_RESULT to truncate_chars."""
        result: list[Message] = []
        protected = self._protected_steps(messages)

        for i, m in enumerate(messages):
            if i == 0 or i == 1:
                result.append(m)
                continue

            if self._is_protected(m, protected):
                result.append(m)
                continue

            if _is_nudge_type(m.meta.type):
                continue

            if m.meta.type == MessageType.TOOL_RESULT:
                if len(m.content) > self.truncate_chars:
                    truncated = Message(
                        role=m.role,
                        content=m.content[: self.truncate_chars],
                        tool_call_id=m.tool_call_id,
                        tool_calls=m.tool_calls,
                        meta=m.meta,
                    )
                    result.append(truncated)
                    continue

            result.append(m)

        return result

    def _phase2_compact(
        self, messages: list[Message], eligible_end: int, counter: TokenCounter
    ) -> list[Message]:
        """Phase 2: phase1 + drop TOOL_RESULT entirely."""
        result: list[Message] = []
        protected = self._protected_steps(messages)

        for i, m in enumerate(messages):
            if i == 0 or i == 1:
                result.append(m)
                continue

            if self._is_protected(m, protected):
                result.append(m)
                continue

            if m.meta.type == MessageType.TOOL_RESULT:
                continue

            result.append(m)

        return result

    def _phase3_compact(
        self, messages: list[Message], eligible_end: int, counter: TokenCounter
    ) -> list[Message]:
        """Phase 3: phase2 + drop REASONING and TEXT_RESPONSE."""
        result: list[Message] = []
        protected = self._protected_steps(messages)

        for i, m in enumerate(messages):
            if i == 0 or i == 1:
                result.append(m)
                continue

            if self._is_protected(m, protected):
                result.append(m)
                continue

            if m.meta.type in (MessageType.REASONING, MessageType.TEXT_RESPONSE):
                continue

            result.append(m)

        return result


def _find_eligible_end(messages: list[Message], keep_recent: int) -> int:
    """Find the last index that is eligible for compaction based on step_index."""
    if keep_recent <= 0:
        return len(messages) - 1

    max_step = -1
    for m in messages:
        if m.meta.step_index is not None and m.meta.step_index > max_step:
            max_step = m.meta.step_index

    if max_step < 0:
        protected_count = keep_recent
        if protected_count >= len(messages):
            protected_count = len(messages)
        return len(messages) - 1 - protected_count

    protected_steps: dict[int, bool] = {}
    current_step = max_step
    for _ in range(keep_recent):
        protected_steps[current_step] = True
        current_step -= 1
        if current_step < 0:
            break

    for i in range(len(messages) - 1, -1, -1):
        step_idx = messages[i].meta.step_index
        if step_idx is None or not protected_steps.get(step_idx, False):
            return i

    return -1

File: typescript/src/llmcontext/strategies.ts
import { Message } from "../llm/request.js";
import { MessageType } from "../middleware/types.js";

export type TokenCounter = (messages: Message[]) => number;

export interface CompactionStrategy {
  compact(
    messages: Message[],
    targetTokens: number,
    counter: TokenCounter
  ): Message[];
  name(): string;
}

function isNudgeType(msgType: MessageType): boolean {
  return (
    msgType === MessageType.STEP_NUDGE ||
    msgType === MessageType.RETRY_NUDGE ||
    msgType === MessageType.PREREQUISITE_NUDGE
  );
}

export class TieredCompact implements CompactionStrategy {
  private keepRecent: number;
  private truncateChars: number;
  public lastPhase: number = 0;

  constructor(keepRecent = 2, truncateChars = 200) {
    this.keepRecent = keepRecent;
    this.truncateChars = truncateChars;
  }

  name(): string {
    return "TieredCompact";
  }

  compact(
    messages: Message[],
    targetTokens: number,
    counter: TokenCounter
  ): Message[] {
    this.lastPhase = 0;

    if (!messages || messages.length === 0) {
      return [];
    }

    const result = [...messages];

    const eligibleEnd = findEligibleEnd(result, this.keepRecent);

    let currentTokens = counter(result);

    if (currentTokens <= targetTokens) {
      return result;
    }

    const phase1Result = this.phase1Compact(result, eligibleEnd, counter);
    currentTokens = counter(phase1Result);
    this.lastPhase = 1;
    if (currentTokens <= targetTokens) {
      return phase1Result;
    }

    const phase2Result = this.phase2Compact(phase1Result, eligibleEnd, counter);
    currentTokens = counter(phase2Result);
    this.lastPhase = 2;
    if (currentTokens <= targetTokens) {
      return phase2Result;
    }

    const phase3Result = this.phase3Compact(phase2Result, eligibleEnd, counter);
    this.lastPhase = 3;
    return phase3Result;
  }

  private protectedSteps(
    messages: Message[]
  ): Map<number, boolean> {
    const protectedSteps = new Map<number, boolean>();

    if (this.keepRecent <= 0) {
      return protectedSteps;
    }

    const steps: number[] = [];
    const stepSet = new Set<number>();

    for (const m of messages) {
      if (m.meta?.step_index !== undefined && !stepSet.has(m.meta.step_index)) {
        steps.push(m.meta.step_index);
        stepSet.add(m.meta.step_index);
      }
    }

    if (steps.length === 0) {
      return protectedSteps;
    }

    if (this.keepRecent >= steps.length) {
      for (const s of steps) {
        protectedSteps.set(s, true);
      }
      return protectedSteps;
    }

    const startIdx = steps.length - this.keepRecent;
    for (let i = startIdx; i < steps.length; i++) {
      protectedSteps.set(steps[i], true);
    }

    return protectedSteps;
  }

  private isProtected(
    msg: Message,
    protectedSteps: Map<number, boolean>
  ): boolean {
    if (msg.meta?.step_index === undefined) {
      return false;
    }
    return protectedSteps.get(msg.meta.step_index) ?? false;
  }

  private phase1Compact(
    messages: Message[],
    _eligibleEnd: number,
    _counter: TokenCounter
  ): Message[] {
    const result: Message[] = [];
    const protectedSteps = this.protectedSteps(messages);

    for (let i = 0; i < messages.length; i++) {
      const m = messages[i];

      if (i === 0 || i === 1) {
        result.push(m);
        continue;
      }

      if (this.isProtected(m, protectedSteps)) {
        result.push(m);
        continue;
      }

      if (isNudgeType(m.meta?.type ?? MessageType.USER_INPUT)) {
        continue;
      }

      if (m.meta?.type === MessageType.TOOL_RESULT) {
        if (m.content.length > this.truncateChars) {
          result.push({
            ...m,
            content: m.content.slice(0, this.truncateChars),
          });
          continue;
        }
      }

      result.push(m);
    }

    return result;
  }

  private phase2Compact(
    messages: Message[],
    _eligibleEnd: number,
    _counter: TokenCounter
  ): Message[] {
    const result: Message[] = [];
    const protectedSteps = this.protectedSteps(messages);

    for (let i = 0; i < messages.length; i++) {
      const m = messages[i];

      if (i === 0 || i === 1) {
        result.push(m);
        continue;
      }

      if (this.isProtected(m, protectedSteps)) {
        result.push(m);
        continue;
      }

      if (m.meta?.type === MessageType.TOOL_RESULT) {
        continue;
      }

      result.push(m);
    }

    return result;
  }

  private phase3Compact(
    messages: Message[],
    _eligibleEnd: number,
    _counter: TokenCounter
  ): Message[] {
    const result: Message[] = [];
    const protectedSteps = this.protectedSteps(messages);

    for (let i = 0; i < messages.length; i++) {
      const m = messages[i];

      if (i === 0 || i === 1) {
        result.push(m);
        continue;
      }

      if (this.isProtected(m, protectedSteps)) {
        result.push(m);
        continue;
      }

      const msgType = m.meta?.type ?? MessageType.TEXT_RESPONSE;
      if (msgType === MessageType.REASONING || msgType === MessageType.TEXT_RESPONSE) {
        continue;
      }

      result.push(m);
    }

    return result;
  }
}

export function findEligibleEnd(
  messages: Message[],
  keepRecent: number
): number {
  if (keepRecent <= 0) {
    return messages.length - 1;
  }

  let maxStep = -1;
  for (const m of messages) {
    if (m.meta?.step_index !== undefined && m.meta.step_index > maxStep) {
      maxStep = m.meta.step_index;
    }
  }

  if (maxStep < 0) {
    const protectedCount = Math.min(keepRecent, messages.length);
    return messages.length - 1 - protectedCount;
  }

  const protectedSteps = new Map<number, boolean>();
  let currentStep = maxStep;
  for (let i = 0; i < keepRecent; i++) {
    protectedSteps.set(currentStep, true);
    currentStep--;
    if (currentStep < 0) {
      break;
    }
  }

  for (let i = messages.length - 1; i >= 0; i--) {
    const stepIdx = messages[i].meta?.step_index;
    if (stepIdx === undefined || !protectedSteps.get(stepIdx)) {
      return i;
    }
  }

  return -1;
}